### PR TITLE
Suggest route optimization on delivery form

### LINKED ISSUE #4432: Suggest users to optimize multi-point deliveries
When a user inputs a delivery with multiple points, we run an optimization in the background, we compare the distance of the optimized route to the one entered by the user, and if it's lower, we suggest the user to accept the suggested route.

### PR DESCRIPTION
**Pull request by alexsegura**
Fixes #4432

suggest_optim

### CODE REVIEW

File: src/Entity/Task.php [hunk 1]
```diff
@@ -1192,4 +1194,13 @@ public function setPrefetchedPackagesAndWeight($prefetchedPackagesAndWeight)
 
         return $this;
     }
+
+    public static function fixTimeWindow(Task $task)
```
Comment (Atala) [inline src/Entity/Task.php:1198]: can we change the name? it is hard to tell what this function is doing

`presetDoneAfter` ?

File: src/Action/Delivery/SuggestOptimizations.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+use AppBundle\Service\RoutingInterface;
+use AppBundle\Vroom\RoutingProblem;
+use AppBundle\Vroom\Vehicle;
+
+class SuggestOptimizations
+{
+    public function __construct(private RouteOptimizer $optimizer, private RoutingInterface $routing)
+    {}
+
+    public function __invoke(Delivery $data): OptimizationSuggestions
```
Comment (Atala) [inline src/Action/Delivery/SuggestOptimizations.php:19]: why not use `RouteOptimizer.optimizePickupsAndDelivery` ? it seems it does more or less the same thing except storing the gain

File: js/app/delivery/form.js [hunk 3]
```diff
@@ -132,6 +135,37 @@ form = new DeliveryForm('delivery', {
...
+              headers: {
+                Accept: 'application/ld+json',
+                'Content-Type': 'application/ld+json',
+                Authorization: `Bearer ${result.jwt}`
+              }
+            })
+            .then(response => {
+              if (response.data.suggestions.length > 0) {
+                this.showSuggestions(response.data.suggestions)
+                // this.reorder(response.data.suggestions[0].order)
```
Comment (Atala) [inline js/app/delivery/form.js:162]: to be removed

**Review by Atala — CHANGES_REQUESTED** [review]
see the comments, it may be useful to use existing "optimizePickupsAndDelivery" function to do the optimization as it looks a lot the same. or maybe move at least the code in the optimizer.

I have found the following bug :
- enter a delivery with 4 points that is suboptimized
- click "yes" in optimization modal
- the points are not in the right order in the form

image

File: js/app/delivery/form/SuggestionModal.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            { suggestedTasks.map((task, index) =>
+              <Timeline.Item key={ `suggested-order-${index}` }>{ task.address?.streetAddress }</Timeline.Item>
+            ) }
+          </Timeline>
+        </div>
+      </div>
+      <div className="d-flex justify-content-between align-items-center">
+        <span>{ t('DELIVERY_OPTIMIZATION_SUGGESTION_CONFIRM') }</span>
+        <div>
+          <button className="btn btn-primary mr-1" type="button" onClick={ onClickYes }>
```
Comment (Atala) [inline js/app/delivery/form/SuggestionModal.js:40]: you should witch yex/no button (no/cancel should be on the left of the modal footer)

File: src/Action/Delivery/SuggestOptimizations.php (hunk 2 above)
Comment (alexsegura) [reply src/Action/Delivery/SuggestOptimizations.php:19]: It doesn't do the same: it's for multiple pickups + one dropoff.

Comment (Atala) [reply src/Action/Delivery/SuggestOptimizations.php:19]: OK
maybe create a new method in the RouteOptimizer ? so we have everything there

**Comment by Paul-Eraman-CoopCycle** [comment]
TO DO

- [ ] try out on LCR

**Comment by Atala** [comment]
- the fix for the failing test is in master I think

- by adding/removing dropoffs I managed to trigger the following error :

**Comment by alexsegura** [comment]
> by adding/removing dropoffs I managed to trigger the following error :
> Uncaught TypeError: can't access property "toLowerCase", task.type is undefined

Ok I see. It happens when doing this:
- Add a new dropoff
- Remove the first one
- Try to edit a property of the added dropoff

It's because of index changes. Looking for a fix.

**Pull request merged**